sdk/search/azure-search-index/design.md

```diff
@@ -0,0 +1,483 @@
+
+In the generated Python API this is reflected in separate methods pairs such as `suggest_get` and `suggestpost`. This GET/POST dichotomy may be unfamiliar or confusing to many users, and does not seem like a useful distinction to directly expose in the user-facing API. Options incude:
+
+* Attempt to automatically choose the variant to use based on the length of the generated query string.
+* Always only use one of the variants (e.g. POST since it can accommodate longer queries).
+* Expose two explicit variants through a flag, using terminology that may be more semantically meaningful to users, e.g.
+  ```python
+  client.get_search_results(search_text="...", mode="simple")   # GET
+  client.get_search_results(search_text="...", mode="full")     # POST
+  ```
+
+***NOTE: Response from Bruce on this matter: "always use POST"***
+
+
+### `SearchApiKeyCredential` API
+
+The Azure Search service uses an HTTP header `"api-key"` to authenticate. This class abstracts the credential to provide room to support other methods of authentication that may become available in the future.
+
+```python
+azure.search.SearchApiKeyCredential(api_key)
+
+# read only property
+SearchAPIKeyCredential.api_key: str
+```
+
+### `IndexOperationBatch` API
+
+Azure Search Service affords the ability to batch CRUD-type operations, returning list of status results corresponding to each individual operation. This class represents a batch of CRUD operations to perfom on an Azure search index. The order of operations is maintained according to the order that a user invokes the methods on an instance.
+
+*NOTE: The acceptibility of alternate CRUD verbs, specific to Azure Search, was discussed in the ADP Arch Board meeting **Azure Search for Java - 10/2/2019***
+
+```python
+azure.search.IndexOperationBatch()
+
+def IndexOperationBatch.add_upload_documents(documents: List[dict])
+
+def IndexOperationBatch.add_delete_documents(documents: List[dict])
+
+def IndexOperationBatch.add_merge_documents(documents: List[dict])
+
+def IndexOperationBatch.add_merge_or_upload_documents(documents: List[dict])
+```
+
+### `SearchIndexClient` API
+
+The search index client is the primary interface for users to interact with an an Azure Search index from the Python API. It accepts credentials and information about the index on creation, and offers methods for all service REST operations.
+
+```python
+azure.search.SearchIndexClient(
+    search_service_name: str,
+    index_name: str,
+    credential: SearchApiKeyCredential
+)
+
+class SearchIndexClient:
+
+    def SearchIndexClient.get_document_count(self) -> int:
+
+    def SearchIndexClient.get_document(
+        self,
+        key: str,
+        selected_fields: List[str] = None
+    ) -> dict:
```

Comment:
Can the output of this method be used as input into the update/merge/mutate operations? Or do I need to transform it somehow? What about using it in a batch operation?

Reply:
There was question in the Java API review about the utility of a "Document" class in this context. I figure that is even moreso in Python. Here documents are JSON dicts and AFAIK as such I would expect they should be ingestible directly by the CRUD operations that also accept JSON dict documents.

Reply:
Documents should be round-trippable, regardless of client library language.

It's important to clearly specify via [tests](https://github.com/Azure/azure-sdk-for-net/blob/90afb070cd980d3da97aaa9227ba67a7e57300e2/sdk/search/Microsoft.Azure.Search/tests/Tests/LookupTests.cs#L23) and [documentation](https://github.com/Azure/azure-sdk-for-net/blob/90afb070cd980d3da97aaa9227ba67a7e57300e2/sdk/search/Microsoft.Azure.Search.Data/src/Customizations/Documents/IDocumentsOperations.Customization.cs#L229) what the mapping is from [Azure Cognitive Search types](https://docs.microsoft.com/rest/api/searchservice/supported-data-types) to Python types (and ditto for every client library language).

Reply:
Good. The reason that I'm asking is that there are some services (and client libraries) that have done some form of transform of the data (it is a flat set of key/values in the response, but a complext object in the request) that end up forcing the user to do unnescessary data transformation/copying of data. Which can range from just being annoying to straight up painful.

Reply:
@johanste Ugh, that sounds painful. Fortunately Azure Cognitive Search doesn't do that. :)

Reply:
I just want to confirm my understanding. In the prototype I have, this works:
```
doc = client.get_document("2")
client.delete_documents({"HotelId": "2"})
r2 = client.upload_documents(doc)
```
where `doc` is a plain python `dict`. @brjohnstmsft it is really the `additional_properties` field you mentioned in an arch board meeting as being what is "really" the document. Is this reasonable/acceptable?

Reply:
@bryevdv In terms of choice of data structure, yes, a dict seems acceptable. However, we have to be careful about what properties end up in the dict. While reviewing the Java SDK, I discovered what appears to be a codegen bug whereby a property defined alongside "additionalProperties" in the Swagger was nevertheless ending up in the dictionary (it's called `@search.score`) during deserialization.

In general we need to be careful that no control information (represented as OData annotations starting with `@`) ends up in the dict that represents the document itself.

Reply:
@brjohnstmsft Thank you, after checking I can see that e.g. `'@odata.context'` is currently being returned. As long as there is a list or decidable criteria (such as "starts with `@odata`), it should be simple to make sure these values do not leak. I will add a note in the "Considerations" section above about this.

Reply:
@bryevdv Anything starting with `@` shouldn't be there, but there are different ways to deal with it:
  - For operations on `SearchIndexClient`, every request should set the `Accept` header to `application/json;odata.metadata=none`. That will ensure that `@odata.context` is not included in the response. In contrast, `SearchServiceClient` should set `Accept` to `application/json;odata.metadata=minimal` to avoid suppressing `@odata.type` annotations in the response (required for deserializing polymorphic model types), but that client doesn't have the same issues around representing documents (since documents are exclusively the domain of `SearchIndexClient`).
  - For any other annotation, they ought to be modeled in the Swagger as separate properties (true today for 2019-05-06), and the code generator should respect that by not putting them in `additionalProperties`.

Reply:
Confirmed adding `{"Accept": "application/json;odata.metadata=none"}` removes the errant fields in the responses. I will update the note above.